frontend/src/lib/interdiff.ts

```diff
@@ -0,0 +1,101 @@
+import { diff as myersDiff } from "fast-myers-diff"
+
+import { DiffOutput, DiffRow } from "./api"
+
+type Chunk = {
+    unaligned: DiffRow[]
+    aligned: DiffRow
+}
+
+function chunkDiffLines(rows: DiffRow[]): { chunks: Chunk[], lastUnaligned: DiffRow[] } {
+    let unaligned = []
+    const chunks = []
+    for (const row of rows) {
+        if (row.base) {
+            chunks.push({ unaligned, aligned: row })
+            unaligned = []
+        } else {
+            unaligned.push(row)
+        }
+    }
+    return { chunks, lastUnaligned: unaligned }
+}
+
+/**
+ * Combine two normal diffs into a three-way diff. This algorithm is derived from asm-differ:
+ * https://github.com/simonlindholm/asm-differ/blob/3841240be5e59c63f735f3cffc5a4c8dd16f14b1/diff.py#L3413-L3447
+ */
+export function interdiff(curr: DiffOutput | null, prev: DiffOutput | null): DiffOutput | null {
+    if (!curr || !prev)
+        return curr
+
+    const rows: DiffRow[] = []
+    const addMatching = (c: DiffRow, p: DiffRow) => {
+        if (c.key === p.key) {
+            rows.push(c)
+        } else {
+            rows.push({
+                key: c.key,
+                base: c.base,
+                current: c.current,
+                previous: p.current,
+            })
+        }
+    }
+
+    const addUnaligned = (cs: DiffRow[], ps: DiffRow[]) => {
+        if (!cs.length && !ps.length)
+            return
+        const ckeys = cs.map(c => c.key)
+        const pkeys = ps.map(p => p.key)
+        let ci = 0, pi = 0
+        // Array.from to silence an error about "Type 'IterableIterator<...>'
+        // can only be iterated through when using the '--downlevelIteration'
+        // flag or with a '--target' of 'es2015' or higher" -- changing
+        // tsconfig compilerOptions.target does not seem to work for me.
+        for (const [c0, c1, p0, p1] of Array.from(myersDiff(ckeys, pkeys))) {
+            while (ci != c0)
+                addMatching(cs[ci++], ps[pi++])
+            while (ci != c1) {
+                const c = cs[ci++]
+                rows.push({
+                    key: c.key,
+                    current: c.current,
+                })
+            }
+            while (pi != p1) {
+                const p = ps[pi++]
+                rows.push({
+                    key: p.key,
+                    previous: p.current,
+                })
+            }
+        }
+        while (ci != cs.length)
+            addMatching(cs[ci++], ps[pi++])
+    }
+
+    const currChunks = chunkDiffLines(curr.rows)
+    const prevChunks = chunkDiffLines(prev.rows)
+    if (currChunks.chunks.length !== prevChunks.chunks.length) {
+        // This should logically never happen, since the two diffs are based on
+        // the same target.
```

Comment:
Warn?

Reply:
Good point. I can add a console.warn; were you thinking some visual indication, and if so do you have tips on how to do it?

Reply:
Just a console warn is fine :shipit:

Reply:
I added a console.warn with the two diff objs being logged to the console; I think that's enough to troubleshoot problems if they come up (I don't expect them to, haven't had reports on asm-differ). This condition will become user-triggerable if we ever make target asm editable, btw, so I don't want to make it too hard an error.